process-migration-service/src/main/java/org/kie/processmigration/service/impl/MigrationServiceImpl.java

```diff
@@ -0,0 +1,200 @@
+        Plan plan = planService.get(definition.getPlanId());
+        if (plan == null) {
+            throw new PlanNotFoundException(definition.getPlanId());
+        }
+        Migration migration = new Migration(definition);
+        em.persist(migration);
+        if (ExecutionType.SYNC.equals(definition.getExecution().getType())) {
+            credentialsService.save(credentials.setMigrationId(migration.getId()));
+            migrate(migration, plan, credentials);
+        } else {
+            schedulerService.scheduleMigration(migration, credentials);
+        }
+        return migration;
+    }
+
+    @Override
+    @Transactional
+    public Migration delete(Long id) {
+        Migration migration = get(id);
+        if (migration == null) {
+            return null;
+        }
+        em.remove(migration);
+        return migration;
+    }
+
+    @Override
+    @Transactional
+    public Migration update(Long id, MigrationDefinition definition, Credentials credentials) throws ReScheduleException {
+        Migration migration = get(id);
+        if (migration == null) {
+            return null;
+        }
+        if (!ExecutionStatus.SCHEDULED.equals(migration.getStatus())) {
+            throw new ReScheduleException("The migration is already started and cannot be re-scheduled");
+        }
+        if (ExecutionType.SYNC.equals(definition.getExecution().getType())) {
+            throw new ReScheduleException("The migration execution type MUST be ASYNC");
+        }
+        migration.setDefinition(definition);
+        em.persist(migration);
+        schedulerService.reScheduleMigration(migration, credentials);
+        return migration;
+    }
+
+    @Override
+    @Transactional
+    public Migration migrate(Migration migration) {
+        Plan plan = planService.get(migration.getDefinition().getPlanId());
+        if (plan == null) {
+            throw new PlanNotFoundException(migration.getDefinition().getPlanId());
+        }
+        Credentials credentials = credentialsService.get(migration.getId());
+        return migrate(migration, plan, credentials);
+    }
+
+    private Migration migrate(Migration migration, Plan plan, Credentials credentials) {
+        try {
+            migration = em.find(Migration.class, migration.getId());
+            em.persist(migration.start());
+            if (credentials == null) {
+                throw new CredentialNotFoundException();
+            }
+            AtomicBoolean hasErrors = new AtomicBoolean(false);
+            //each instance id will spawn its own request to KIE server for migration.
+            List<Long> instancesIdList = migration.getDefinition().getProcessInstanceIds();
+            for (Long instanceId : instancesIdList) {
+                MigrationReportInstance reportInstance = kieService
+                                                                   .createProcessAdminServicesClient(CredentialsProviderFactory.getProvider(credentials))
```

Comment:
since the work ins done in a synchronous loop (no multi threading) I'd create ProcessAdminServicesClient once and reuse it